Convert Multus to the Sidecar pattern

mateoflorido opened the pull request:
## Changes
- Convert the Charm from PodSpec pattern to the Sidecar pattern.

*IMPORTANT: Tests fail in the teardown phase due to a bug in the Juju event handler for Sidecar charms. The `remove` event is not fired, and therefore the resources are not completely cleaned up. Relevant link: [LP#1951415](https://bugs.launchpad.net/juju/+bug/1951415)*

canonicalmateo committed:
Conversion to Sidecar pattern

canonicalmateo committed:
Fix lint

canonicalmateo committed:
Remove cached-property

Cynerva commented on src/charm.py:
```diff
@@ -1,180 +1,95 @@
+
+    def _update_status(self, _):
+        if not self.stored.deployed:
+            return
```
This generally will not be true for non-leader units, because of the early return in `_install_or_upgrade`, right? Will they just sit in Unknown status forever? I think that could cause problems for the `juju-wait` CLI tool and similar "wait until the model is ready" functions, so I'd probably just put them in Active status if they're not going to do anything.

Cynerva commented on src/charm.py:
```diff
@@ -1,180 +1,95 @@
+    def _on_config_changed(self, _):
+        na_definitions = self.config.get("network-attachment-definitions")
+        if na_definitions:
+            self.nad_manager.apply_manifests(na_definitions)
```
What if the `release` config changes? Would we need to re-apply Multus manifests here?

Cynerva commented on src/charm.py:
```diff
@@ -1,180 +1,95 @@
+        log.info("Removing Network Attachment Definitions")
+        self.nad_manager.remove_resources()
+        log.info("Removing Multus manifests")
+        self.manifests.delete_manifests(ignore_unauthorized=True, ignore_not_found=True)
```
I can't confirm this right now because of the upstream bug, but based on the remove event description [here](https://juju.is/docs/sdk/remove-event), it sounds like the remove event will run on every unit that gets removed. If that's true, then something like:

```
juju add-unit multus
juju remove-unit multus/1
```

could remove Multus and its resources, even though the application still exists with a running unit in Juju. I don't think that's quite what we want.

It's something to investigate further once the Juju bug gets fixed, I suppose.

Cynerva commented on src/manifests.py:
```diff
@@ -0,0 +1,25 @@
+class MultusManifests(Manifests):
+    def __init__(self, charm, charm_config):
+        manipulations = [
+            ManifestLabel(self),
```
Should the ConfigRegistry manipulation be in here, too? Without it, the DaemonSet will be applied with an unmodified image field that pulls from ghcr.io, right?

We need the Multus image to come from rocks.canonical.com/cdk by default. I recommend adding an image-registry charm config option to control this.

Cynerva commented on src/net_attach_definitions.py:
```diff
@@ -0,0 +1,133 @@
+            with open("schemas/NetworkAttachDefinition.yaml", "r") as f:
+                return yaml.safe_load(f)
+        except yaml.YAMLError:
+            log.error("Failed reading validation schema")
```
I think it would be good to log more here - what was the YAMLError exception's message string?

I'm personally a fan of dumping the whole exception with `log.error(traceback.format_exc())`, although I think some people might find that to be too verbose. I'd be happy to at least see the exception's message string.

Cynerva commented on src/net_attach_definitions.py:
```diff
@@ -0,0 +1,133 @@
+                self.client.apply(rsc.resource, force=True)
+                applied.add(rsc)
+            except ApiError:
+                log.exception(f"Failed applying {rsc}")
```
Can we log details about the ApiError before we drop it?

Cynerva replied:
Also, I think we should inform the user that we failed and will retry. Waiting status might be appropriate here.

Actually, will we retry? I think we should - this could be caused by a temporary outage of the Kubernetes API, right?

Cynerva commented on src/net_attach_definitions.py:
```diff
@@ -0,0 +1,133 @@
+            resources = self._validate_and_load(manifests)
+        except (self.ValidationError, yaml.YAMLError) as e:
+            log.error(e)
+            return
```
If this fails, I think we should inform the user in _update_status by putting the charm in Blocked status with a helpful message, or something like that.

Cynerva commented on src/net_attach_definitions.py:
```diff
@@ -0,0 +1,133 @@
+            log.error(
+                "Failed to get Network Attachment Definitions in cluster. "
+                "There may be NADs that couldn't be removed."
+            )
```
Can we log the ApiError details here?

Cynerva commented on src/net_attach_definitions.py:
```diff
@@ -0,0 +1,133 @@
+                "Failed to get Network Attachment Definitions in cluster. "
+                "There may be old NADs that couldn't be removed. "
+                "Try again later running the scrub_nads action."
+            )
```
Can we log the ApiError details here?

Cynerva commented on src/net_attach_definitions.py:
```diff
@@ -0,0 +1,133 @@
+            return
+        remnants = installed.difference(self.resources)
+        for r in remnants:
+            self.client.delete(type(r.resource), r.name, namespace=r.namespace)
```
Hmm, is this going to delete every NAD in the cluster that the charm doesn't recognize? What if the user manually created NADs with `kubectl`? Will those get deleted too?

I would be careful not to step on the user experience for NetworkAttachmentDefinitions created via kubectl. The way I handled this before was by having the charm [label every NAD it creates](https://github.com/charmed-kubernetes/charm-multus/blob/15d880711bc117a49c187e49acda01b7316ff53e/net-attach-def-manager/net-attach-def-manager.py#L23) and [only scrub NADs that have that label](https://github.com/charmed-kubernetes/charm-multus/blob/15d880711bc117a49c187e49acda01b7316ff53e/net-attach-def-manager/net-attach-def-manager.py#L41). I think it would be good to do something similar here.

Cynerva reviewed (changes requested):
This is looking good, nice work. I do have some requests and questions, see comments.

canonicalmateo committed:
Address comments

canonicalmateo committed:
Fix unit tests

mateoflorido commented:
Thanks @Cynerva  for the review. The last commits addressed the previous review.
## Changes
- Non-leader units are set to active status by default.
- Added `image-registry` config option, defaults to `rocks.canonical.com`.
- Increased log verbosity.
- Invalid Network Attachment Definitions YAML now puts the charm in a Blocked State.
- Network Attachment Definitions applied by the charm are now labeled.
- Added `retry` in case the requests fail due to  ApiError exception.

kwmonroe reviewed (approved):
LGTM; i especially like how you centrally handled blocked vs waiting (retrying) event.
